### PR TITLE
Coffee machines carry over reagents from the beans

### PR DESCRIPTION
**Pull request by disappointedButNotSuprised**
## About The Pull Request

I can't believe that this wasn't included in the original PR #70991. 
From now on, if there might be any unusual substances inside coffee beans, their reagents will be carried over to the coffee made by the machine. This is one of those niche unused mechanics that no one knows about but the coders hope to see one day in a intricate plot situation.
<sub><sup>please please please I hope so much for someone to use this</sup></sub>

Obviously a indispensable gampleay feature.

https://github.com/tgstation/tgstation/assets/57324037/e3c243d7-03fd-47bc-9bdb-fac50c5bba0b

## Why It's Good For The Game

Provides the coffee machine with features that it should have had fromm the beginning.

## Changelog

:cl:
add: Coffee machines can now carry over reagents from the beans to the coffee (surely no one will inject poison into them)
/:cl:

### CODE REVIEW

**Comment by san7890** [comment]
can you elaborate on this change in the changelog entry

**Comment by optimumtact** [comment]
this will just mean people dont drink coffee it's over

**Comment by WarlockD** [comment]
It just means command trusts the coffee from pods, just never from botany.

**Comment by SomeoneYouProbablyKnow** [comment]
Good, no one should trust botany with anything

**Comment by EricZilla** [comment]
Strange Beagents

**Comment by DaCoolBoss** [comment]
Don't talk to me until I've had my ephedrine laced Robusta coffee

File: code/modules/food_and_drinks/machinery/coffeemaker.dm [hunk 1]
```diff
@@ -715,9 +715,30 @@
...
 	operate_for(brew_time)
-	coffeepot.reagents.add_reagent_list(list(/datum/reagent/consumable/coffee = 120))
-	coffee.Cut(1,2) //remove the first item from the list
+
+	// create a reference bean reagent list
+	var/list/reference_bean_reagents = list()
+	var/obj/item/food/grown/coffee/reference_bean = new /obj/item/food/grown/coffee(src)
+	for(var/datum/reagent/ref_bean_reagent in reference_bean.reagents.reagent_list)
+		reference_bean_reagents += ref_bean_reagent.name
+
```
Comment (SyncIt21) [inline code/modules/food_and_drinks/machinery/coffeemaker.dm:724]:
```suggestion
qdel(reference_bean)
```
Dispose the reference bean after your done with it.

File: code/modules/food_and_drinks/machinery/coffeemaker.dm [hunk 2]
```diff
@@ -715,9 +715,30 @@
 	if(!try_brew())
 		return
 	operate_for(brew_time)
-	coffeepot.reagents.add_reagent_list(list(/datum/reagent/consumable/coffee = 120))
-	coffee.Cut(1,2) //remove the first item from the list
+
+	// create a reference bean reagent list
+	var/list/reference_bean_reagents = list()
+	var/obj/item/food/grown/coffee/reference_bean = new /obj/item/food/grown/coffee(src)
+	for(var/datum/reagent/ref_bean_reagent in reference_bean.reagents.reagent_list)
```
Comment (SyncIt21) [inline code/modules/food_and_drinks/machinery/coffeemaker.dm:722]:
```suggestion
	for(var/datum/reagent/ref_bean_reagent as anything in reference_bean.reagents.reagent_list)
```

File: code/modules/food_and_drinks/machinery/coffeemaker.dm [hunk 3]
```diff
@@ -715,9 +715,30 @@
...
+		reference_bean_reagents += ref_bean_reagent.name
+
+	// add all the reagents from the coffee beans to the coffeepot (ommit the ones from the reference bean)
+	var/list/reagent_delta = list()
+	var/obj/item/food/grown/coffee/bean = coffee[coffee_amount]
+	for(var/datum/reagent/substance in bean.reagents.reagent_list)
```
Comment (SyncIt21) [inline code/modules/food_and_drinks/machinery/coffeemaker.dm:728]:
```suggestion
	for(var/datum/reagent/substance as anything in bean.reagents.reagent_list)
```

File: code/modules/food_and_drinks/machinery/coffeemaker.dm [hunk 4]
```diff
@@ -715,9 +715,30 @@
...
+		if(!(reference_bean_reagents.Find(substance.name)))	// we only add the reagent if it's a non-standard for coffee beans
```
Comment (SyncIt21) [inline code/modules/food_and_drinks/machinery/coffeemaker.dm:729]: Why do we not want to include the reagents in a standard coffee bean? What are the implications?

File: code/modules/food_and_drinks/machinery/coffeemaker.dm [hunk 5]
```diff
@@ -715,9 +715,30 @@
...
+	coffeepot.reagents.add_reagent_list(reagent_delta)
+
+	// remove the coffee beans from the machine
+	coffee.Cut(1,2)
 	coffee_amount--
+
+	// fill the rest of the pot with coffee
+	if(coffeepot.reagents.total_volume < 120)
+		var/extra_coffee_amount = 120 - coffeepot.reagents.total_volume
+		coffeepot.reagents.add_reagent_list(list(/datum/reagent/consumable/coffee = extra_coffee_amount))
```
Comment (SyncIt21) [inline code/modules/food_and_drinks/machinery/coffeemaker.dm:740]:
```suggestion
		coffeepot.reagents.add_reagent(/datum/reagent/consumable/coffee, extra_coffee_amount)
```
If you are adding only 1 reagent don't wrap it in a list. Use the proc directly

**Review by SyncIt21 — CHANGES_REQUESTED** [review]

File: code/modules/food_and_drinks/machinery/coffeemaker.dm (hunk 4 above)
Comment (disappointedButNotSuprised) [reply code/modules/food_and_drinks/machinery/coffeemaker.dm:729]: All the coffee obtained elsewhere (vendomats, the normal coffeemaker etc.) does follow this composition (just coffee). I figured it would be best to just keep it that way. Otherwise the coffee would include `Vitamin`, `Nutrient`, and `Coffeepowder` which I think would only add clutter

Comment (SyncIt21) [reply code/modules/food_and_drinks/machinery/coffeemaker.dm:729]: Fair enough. Ideally all the standard reagents inside a coffee bean should be converted into coffee and not omitted like this but there is no clean way of doing that. You can resolve this

**Comment by DATA-xPUNGED** [comment]
> this will just mean people dont drink coffee it's over

If you offer a drink to an ss13 player and say it's poisoned they're gonna drink it anyways

**Comment by Admiral-Gigaton** [comment]
If you put the reagent-filled beans into the machine, will the reagents heat in the coffee? Potential to blow up the bridge assistant is great

**Comment by disappointedButNotSuprised** [comment]
> If you put the reagent-filled beans into the machine, will the reagents heat in the coffee? Potential to blow up the bridge assistant is great

Turns out that although coffee in ss13 heats you up, it is not hot _per se_. Temperature of reagents is typically handled by a property of the reagent container, not the reagent itself, and thus  """hot cofee""" has the `DEFAULT_REAGENT_TEMPERATURE` of the coffeepot equal to 300K (~26 deg C). Reagents cannot be inductively heated by other reagents.

This is however a super funny idea I might implement this in a future PR where all coffee could be made actually _hot_ by raising the default container temps in coffee producing machines (might also consider doing this on other food related machinery if applicable).

TLDR; No

**Comment by Admiral-Gigaton** [comment]
> Turns out that although coffee in ss13 heats you up, it is not hot _per se_. Temperature of reagents is typically handled by a property of the reagent container, not the reagent itself, and thus """hot cofee""" has the `DEFAULT_REAGENT_TEMPERATURE` of the coffeepot equal to 300K (~26 deg C). Reagents cannot be inductively heated by other reagents.
> 
> This is however a super funny idea I might implement this in a future PR where all coffee could be made actually _hot_ by raising the default container temps in coffee producing machines (might also consider doing this on other food related machinery if applicable).
> 
> TLDR; No

Well, if that's the case, can you add some function to be able to change the output temp of the coffee? Not unlike showers, so hacking it (or emagging, if too OP) could allow you to adjust its temp output, or something like screwdrivering it will change its temp.

Of course, not required, would just add more expansion to that idea for the coffee maker

**Comment by jlsnow301** [comment]
coffee buff
